// botfuel-dialog/build/dialogs/cancellation-dialog.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Copyright (c) 2017 - present, Botfuel (https://www.botfuel.io).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var PromptDialog = require('./prompt-dialog');

var CancellationDialog = function (_PromptDialog) {
  _inherits(CancellationDialog, _PromptDialog);

  function CancellationDialog() {
    _classCallCheck(this, CancellationDialog);

    return _possibleConstructorReturn(this, (CancellationDialog.__proto__ || Object.getPrototypeOf(CancellationDialog)).apply(this, arguments));
  }

  _createClass(CancellationDialog, [{
    key: 'dialogWillComplete',
    value: function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(userMessage, data) {
        var answer;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (!(data.missingEntities.size === 0)) {
                  _context.next = 5;
                  break;
                }

                _context.next = 3;
                return this.brain.conversationSet(userMessage.user, this.parameters.namespace, {});

              case 3:
                answer = data.matchedEntities.answer.values[0].value;
                return _context.abrupt('return', answer === false ? this.complete() : this.cancelPrevious());

              case 5:
                return _context.abrupt('return', this.wait());

              case 6:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function dialogWillComplete(_x, _x2) {
        return _ref.apply(this, arguments);
      }

      return dialogWillComplete;
    }()
  }]);

  return CancellationDialog;
}(PromptDialog);

CancellationDialog.params = {
  namespace: 'cancellation',
  entities: {
    answer: {
      dim: 'system:boolean'
    }
  }
};

module.exports = CancellationDialog;
